src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -155,36 +162,85 @@ public async Task<WorkItem> CreateRcaWorkItem(string project, string title, Canc
         return JsonConvert.DeserializeObject<WorkItem>(json);
     }
 
-    public async Task<string> TryGetImageName(
+    /// <summary>
+    /// The method reads the logs as a stream, line by line and tries to match the regexes in order. 
+    /// If the regexes are matched in order, the last match is returned.
+    /// </summary>
+    public async Task<string?> TryGetImageName(
         string logUri,
-        Func<string, string> findImageName,
+        Regex[] regexes,
         CancellationToken cancellationToken)
     {
-
         using var request = new HttpRequestMessage(HttpMethod.Get, logUri);
         request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/plain"));
 
         using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
         using Stream logStream = await response.Content.ReadAsStreamAsync(cancellationToken);
         using StreamReader reader = new StreamReader(logStream);
-        string line;
-        while ((line = reader.ReadLine()) != null)
+        int regexIndex = 0;
+        string? line;
+        Queue<string> lineCache = new Queue<string>();
+        while ((line = await reader.ReadLineAsync()) != null)
         {
-            var imageName = findImageName(line);
-            if (imageName != null)
+            if (TryMatchRegex(line, regexes[regexIndex], out string? result))
             {
-                return imageName;
+                // We need to cache lines to check the regex sequence again in case of a miss.
+                // We don't want to cache the line that was already matched by the first regex
+                if (regexIndex > 0)
+                {
+                    lineCache.Enqueue(line);
+                }
+
+                regexIndex++;
+
+                if (regexIndex == regexes.Length)
+                {
+                    return result;
+                }
+            }
+            else if (regexIndex > 0)
+            {
+                // Add the line that we missed on to the cache, it might fit the 
+                lineCache.Enqueue(line);
+                regexIndex = 0;
+
+                // We missed, time to go back to the cache and look if the sequence there will match
+                while (lineCache.Count > 0)
+                {
+                    var cachedLine = lineCache.Dequeue();
+                    if (TryMatchRegex(cachedLine, regexes[regexIndex], out result))
+                    {
+                        regexIndex++;
```

Comment:
It's not obvious to me whether `lineCache.Count` can be larger than `regexes.Length` because then we would have a problem.